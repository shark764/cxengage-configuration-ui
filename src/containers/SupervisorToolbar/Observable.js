import { fromEvent } from 'rxjs/observable/fromEvent';
import 'rxjs/add/operator/map';
import 'rxjs/add/operator/filter';

import { requestingMonitorCall } from '../../redux/modules/supervisorToolbar';

import store from '../../redux/store.js';

export const filterArray = [
  'cxengage/session/sqs-shut-down',
  'cxengage/twilio/device-ready',
  'cxengage/interactions/voice/silent-monitor-start',
  'cxengage/interactions/voice/silent-monitor-end',
  'cxengage/interactions/voice/unmute-acknowledged',
  'cxengage/interactions/voice/mute-acknowledged',
  'cxengage/session/started',
  'monitorCall'
];

export function messageObservable() {
  return fromEvent(window, 'message')
    .filter(
      ({ data }) =>
        data.subscription && filterArray.includes(data.subscription.topic)
    )
    .map(event => {
      if (event.data.subscription.topic === 'monitorCall') {
        return store.dispatch(
          requestingMonitorCall(event.data.subscription.response.interactionId)
        );
      } else {
        return store.dispatch({
          type: event.data.subscription.topic,
          response: event.data.subscription.response
        });
      }
    });
}

let subscription;
export function subscribe() {
  subscription = messageObservable().subscribe();
}
export function unsubscribe() {
  subscription.unsubscribe();
}
