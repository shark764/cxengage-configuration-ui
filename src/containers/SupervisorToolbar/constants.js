export const TOGGLE_TWILIO_ENABLED = 'src/SupervisorToolbar/SET_TWILIO_ENABLED';
export const TOGGLE_TWILIO_IS_DEFAULT_EXTENSION =
  'src/SupervisorToolbar/TOGGLE_TWILIO_IS_DEFAULT_EXTENSION';
export const SET_SILENT_MONITORING_STATUS =
  'src/SupervisorToolbar/SET_SILENT_MONITORING_STATUS';
export const TOGGLE_MUTE = 'src/SupervisorToolbar/TOGGLE_MUTE';
export const SET_SILENT_MONITORING_INTERACTION_ID =
  'src/SupervisorToolbar/SET_SILENT_MONITORING_INTERACTION_ID';
