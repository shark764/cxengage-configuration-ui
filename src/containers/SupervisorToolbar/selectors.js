import { createSelector } from 'reselect';

const selectSupervisorToolbarMap = state => state.get('SupervisorToolbar');
// const selectInteractionMonitoringMap = state =>
//   state.get("InteractionMonitoring");

export const selectSupervisorToolbar = createSelector(
  selectSupervisorToolbarMap,
  supervisorToolbar => supervisorToolbar
);
