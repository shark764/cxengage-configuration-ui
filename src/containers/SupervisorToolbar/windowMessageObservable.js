import { fromEvent } from 'rxjs/observable/fromEvent';
import 'rxjs/add/operator/map';
import 'rxjs/add/operator/filter';
import 'rxjs/add/operator/do';

import { requestingMonitorCall } from '../../redux/modules/supervisorToolbar';

import store from '../../redux/store.js';

export const filterArray = [
  'cxengage/session/sqs-shut-down',
  'cxengage/twilio/device-ready',
  'cxengage/interactions/voice/silent-monitor-start',
  'cxengage/interactions/voice/silent-monitor-end',
  'cxengage/interactions/voice/unmute-acknowledged',
  'cxengage/interactions/voice/mute-acknowledged',
  'cxengage/session/started',
  'monitorCall'
];

export function messageObservable() {
  return fromEvent(window, 'message')
    .filter(({ data }) => data.subscription && filterArray.includes(data.subscription.topic))
    .map(event => {
      let subscription = event.data.subscription;
      if (subscription.topic === 'monitorCall') {
        store.dispatch(
          requestingMonitorCall(
            subscription.response.interactionId,
            subscription.response.transitionCall,
            subscription.response.chosenExtension
          )
        );
      } else {
        store.dispatch({
          type: subscription.topic,
          response: subscription.response
        });
      }
      return event;
    })
    .do(({ data }) => {
      data.subscription.topic !== 'monitorCall' &&
        localStorage.setItem(
          'SupervisorToolbar',
          JSON.stringify(
            store
              .getState()
              .get('SupervisorToolbar')
              .toJS()
          )
        );
    })
    .map(event => ({ type: `${event.data.subscription.topic}_$` }));
}

let subscription;
export function messageSubscribe() {
  subscription = messageObservable().subscribe();
}
export function messageUnsubscribe() {
  subscription.unsubscribe();
}
