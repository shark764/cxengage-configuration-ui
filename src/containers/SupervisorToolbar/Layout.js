import React, { Component } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { subscribe, unsubscribe } from './Observable.js';
import { MutedIconSVG, UnMutedIconSVG, HangUpIconSVG } from 'cx-ui-components';

const HangUpIconSVGWrapper = styled(HangUpIconSVG)`
  margin-left: 10px;
  margin-right: 10px;
`;
const Toolbar = styled.div`
  position: fixed;
  bottom: 0px;
  padding-bottom: 10px;
`;

export default class SupervisorToolbar extends Component {
  componentWillMount() {
    this.props.startSupervisorToolbarSubscriptions();
    subscribe();
  }

  componentWillUnmount() {
    unsubscribe();
  }

  render() {
    return (
      <div id="SupervisorToolbar">
        {this.props.monitoringStatus !== 'offline' && (
          <Toolbar>
            <HangUpIconSVGWrapper
              id="hangUpButton"
              className="HangUpIconSVG"
              onClick={this.props.requestingHangUp}
              loading={this.props.monitoringStatus !== 'connected'}
              size={40}
            />
            {this.props.monitoringStatus === 'connected' &&
              (this.props.muted ? (
                <MutedIconSVG
                  id="unmuteButton"
                  onClick={this.props.requestingToggleMute}
                  size={40}
                />
              ) : (
                <UnMutedIconSVG
                  id="muteButton"
                  onClick={this.props.requestingToggleMute}
                  size={40}
                />
              ))}
          </Toolbar>
        )}
      </div>
    );
  }
}

SupervisorToolbar.propTypes = {
  startSupervisorToolbarSubscriptions: PropTypes.func.isRequired,
  requestingToggleMute: PropTypes.func.isRequired,
  requestingHangUp: PropTypes.func.isRequired,
  muted: PropTypes.bool.isRequired,
  twilioEnabled: PropTypes.bool.isRequired,
  monitoringStatus: PropTypes.string.isRequired,
  interactionId: PropTypes.string.isRequired
};
