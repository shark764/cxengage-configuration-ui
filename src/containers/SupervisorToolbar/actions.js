import * as ACTIONS from './constants';

export function setTwilioEnabled() {
  return {
    type: ACTIONS.SET_TWILIO_ENABLED
  };
}
