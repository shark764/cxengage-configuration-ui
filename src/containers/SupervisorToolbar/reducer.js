import * as ACTIONS from './constants';
import { fromJS } from 'immutable';

const initialState = fromJS({
  twilio: {
    enabled: false,
    isDefaultExtension: false
  },
  silentMonitoring: {
    status: 'offline',
    interactionId: 'na'
  },
  muted: true
});

function supervisorToolbarReducer(state = initialState, action) {
  switch (action.type) {
    case ACTIONS.TOGGLE_TWILIO_ENABLED:
      return state.setIn(
        ['twilio', 'enabled'],
        !state.getIn(['twilio', 'enabled'])
      );
    case ACTIONS.TOGGLE_TWILIO_IS_DEFAULT_EXTENSION:
      return state.setIn(
        ['twilio', 'isDefaultExtension'],
        !state.getIn(['twilio', 'isDefaultExtension'])
      );
    case ACTIONS.SET_SILENT_MONITORING_STATUS:
      return state.setIn(['silentMonitoring', 'status'], action.status);
    case ACTIONS.SET_SILENT_MONITORING_INTERACTION_ID:
      return state.setIn(
        ['silentMonitoring', 'interactionId'],
        action.interactionId
      );
    case ACTIONS.TOGGLE_MUTE:
      return state.set('muted', !state.get('muted'));
    default:
      return state;
  }
}

export default supervisorToolbarReducer;
