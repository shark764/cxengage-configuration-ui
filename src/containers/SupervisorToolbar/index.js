import { connect } from 'react-redux';

import {
  selectSupervisorToolbarMuted,
  selectSupervisorToolbarTwilioEnabled,
  selectSupervisorToolbarTwilioIsDefaultExtension,
  selectSupervisorToolbarSilentMonitoringStatus,
  selectSupervisorToolbarSilentMonitoringInteractionId
} from '../../redux/modules/supervisorToolbar/selectors';

import {
  startSupervisorToolbarSubscriptions,
  requestingToggleMute,
  requestingHangUp
} from '../../redux/modules/supervisorToolbar';

import { userHasBargeAllCallsPermission } from '../../redux/modules/reporting/interactionMonitoring/selectors';

import Layout from './layout';
import { isInIframe } from 'serenova-js-utils/browser';

export const mapStateToProps = (state, props) => ({
  muted: selectSupervisorToolbarMuted(state, props),
  twilioEnabled: selectSupervisorToolbarTwilioEnabled(state, props),
  twilioIsDefault: selectSupervisorToolbarTwilioIsDefaultExtension(state, props),
  monitoringStatus: selectSupervisorToolbarSilentMonitoringStatus(state, props),
  interactionId: selectSupervisorToolbarSilentMonitoringInteractionId(state, props),
  userHasBargeAllCallsPermission: userHasBargeAllCallsPermission(state),
  insideIframe: !isInIframe()
});

const actions = {
  startSupervisorToolbarSubscriptions,
  requestingToggleMute,
  requestingHangUp
};

export default connect(mapStateToProps, actions)(Layout);
