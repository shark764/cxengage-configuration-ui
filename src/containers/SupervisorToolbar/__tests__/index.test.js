/*
 * Copyright Â© 2015-2018 Serenova, LLC. All rights reserved.
 */

import React from 'react';
import { createStore } from 'redux';
import { shallow } from 'enzyme';
import SupervisorToolbar, { mapStateToProps } from '../';
jest.mock('../../../redux/store.js', () => jest.fn());

// Mock all the required selectors
jest.mock('../../../redux/modules/supervisorToolbar/selectors', () => ({
  selectSupervisorToolbarMuted: () => true,
  selectSupervisorToolbarTwilioEnabled: () => true,
  selectSupervisorToolbarTwilioIsDefaultExtension: () => true,
  selectSupervisorToolbarSilentMonitoringStatus: () => 'conected',
  selectSupervisorToolbarSilentMonitoringInteractionId: () =>
    'mockInteractionId'
}));
jest.mock(
  '../../../redux/modules/reporting/interactionMonitoring/selectors',
  () => ({
    userHasBargeAllCallsPermission: () => true
  })
);

describe('Supervisor Toolbar Renders', () => {
  it('renders', () => {
    const store = createStore(state => state);
    shallow(<SupervisorToolbar store={store}>Child</SupervisorToolbar>);
  });
});

describe('Maps state to props only using selectors', () => {
  it('validates object created from mapStateToProps', () => {
    expect(mapStateToProps()).toMatchSnapshot();
  });
});
