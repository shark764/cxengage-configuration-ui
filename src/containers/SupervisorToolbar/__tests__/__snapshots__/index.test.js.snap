// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Maps state to props only using selectors validates object created from mapStateToProps 1`] = `
Object {
  "interactionId": "mockInteractionId",
  "monitoringStatus": "conected",
  "muted": true,
  "twilioEnabled": true,
  "twilioIsDefault": true,
  "userHasBargeAllCallsPermission": true,
}
`;
